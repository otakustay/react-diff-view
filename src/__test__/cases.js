/* In case you are looking for type of hunk, see src/Hunk/index.js */
import {parseDiff} from '..';

export const basic = `diff --git a/src/__test__/index.test.jsx b/src/__test__/index.test.jsx
index 643c2f0..7883597 100644
--- a/src/__test__/index.test.jsx
+++ b/src/__test__/index.test.jsx
@@ -21,3 +21,3 @@ describe('basic test', () => {
     test('App renders correctly', () => {
-        expect(renderer.create(<App diffText={'deff'} />).toJSON()).toMatchSnapshot();
+        expect(renderer.create(<App diffText={'diff'} />).toJSON()).toMatchSnapshot();
     });`;

export const basicHunk = parseDiff(basic)[0].hunks[0];

export const multiple = `diff --git a/src/index1.js b/src/index1.js
index e69de29..d00491f 100644
--- a/src/index1.js
+++ b/src/index1.js
@@ -0,0 +1 @@
+1
diff --git a/src/index.js b/src/index2.js
index d00491f..0cfbf08 100644
--- a/src/index2.js
+++ b/src/index2.js
@@ -1 +1 @@
-1
+2
diff --git a/src/index3.js b/src/index3.js
index 0cfbf08..e69de29 100644
--- a/src/index3.js
+++ b/src/index3.js
@@ -1 +0,0 @@
-2`;

export const multipleHunk = parseDiff(basic)[0].hunks[0];
