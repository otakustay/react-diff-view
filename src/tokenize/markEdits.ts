// @ts-nocheck ignore: third-party
import {findIndex, flatMap, last, flatten} from 'lodash';
import DiffMatchPatch from 'diff-match-patch';
import pickRanges from './pickRanges';

// 临时挖出来
const DIFF_DELETE = -1;
const DIFF_INSERT = 1;
const DIFF_EQUAL = 0;
// const {DIFF_EQUAL, DIFF_DELETE, DIFF_INSERT} = DiffMatchPatch;

const findChangeBlocks = changes => {
    const start = findIndex(changes, change => !change.isNormal);

    if (start === -1) {
        return [];
    }

    const end = findIndex(changes, change => change.isNormal, start);

    if (end === -1) {
        return [changes.slice(start)];
    }

    return [
        changes.slice(start, end),
        ...findChangeBlocks(changes.slice(end)),
    ];
};

const groupDiffs = diffs => diffs.reduce(
    ([oldDiffs, newDiffs], diff) => {
        const [type] = diff;

        switch (type) {
            case DIFF_INSERT:
                newDiffs.push(diff);
                break;
            case DIFF_DELETE:
                oldDiffs.push(diff);
                break;
            default:
                oldDiffs.push(diff);
                newDiffs.push(diff);
                break;
        }

        return [oldDiffs, newDiffs];
    },
    [[], []]
);

const splitDiffToLines = diffs => diffs.reduce(
    (lines, [type, value]) => {
        const currentLines = value.split('\n');

        const [currentLineRemaining, ...nextLines] = currentLines.map(line => [type, line]);
        const next = [
            ...lines.slice(0, -1),
            [...last(lines), currentLineRemaining],
            ...nextLines.map(line => [line]),
        ];
        return next;
    },
    [[]]
);

const diffsToEdits = (diffs, lineNumber) => {
    const output = diffs.reduce(
        (output, diff) => {
            const [edits, start] = output;
            const [type, value] = diff;
            if (type !== DIFF_EQUAL) {
                const edit = {
                    type: 'edit',
                    lineNumber: lineNumber,
                    start: start,
                    length: value.length,
                };
                edits.push(edit);
            }

            return [edits, start + value.length];
        },
        [[], 0]
    );

    return output[0];
};

const convertToLinesOfEdits = (linesOfDiffs, startLineNumber) => flatMap(
    linesOfDiffs,
    (diffs, i) => diffsToEdits(diffs, startLineNumber + i)
);

const diffText = (x, y) => {
    const dmp = new DiffMatchPatch();
    const diffs = dmp.diff_main(x, y);
    dmp.diff_cleanupSemantic(diffs);

    if (diffs.length <= 1) {
        return [[], []];
    }

    return groupDiffs(diffs);
};

const diffChangeBlock = changes => {
    const [oldSource, newSource] = changes.reduce(
        ([oldSource, newSource], {isDelete, content}) => (
            isDelete
                ? [oldSource + (oldSource ? '\n' : '') + content, newSource]
                : [oldSource, newSource + (newSource ? '\n' : '') + content]
        ),
        ['', '']
    );

    const [oldDiffs, newDiffs] = diffText(oldSource, newSource);

    if (oldDiffs.length === 0 && newDiffs.length === 0) {
        return [oldDiffs, newDiffs];
    }

    const getLineNumber = change => change && change.lineNumber;
    const oldStartLineNumber = getLineNumber(changes.find(change => change.isDelete));
    const newStartLineNumber = getLineNumber(changes.find(change => change.isInsert));
    const oldEdits = convertToLinesOfEdits(splitDiffToLines(oldDiffs), oldStartLineNumber);
    const newEdits = convertToLinesOfEdits(splitDiffToLines(newDiffs), newStartLineNumber);

    return [oldEdits, newEdits];
};

const diffByLine = changes => changes.reduce(
    ([oldEdits, newEdits, previousChange], currentChange) => {
        if (!previousChange.isDelete || !currentChange.isInsert) {
            return [oldEdits, newEdits, currentChange];
        }

        const [oldDiffs, newDiffs] = diffText(previousChange.content, currentChange.content);
        return [
            oldEdits.concat(diffsToEdits(oldDiffs, previousChange.lineNumber)),
            newEdits.concat(diffsToEdits(newDiffs, currentChange.lineNumber)),
            currentChange,
        ];
    },
    [[], [], {}]
);

export default (hunks, {type = 'block'} = {}) => {
    const changeBlocks = flatMap(hunks.map(hunk => hunk.changes), findChangeBlocks);
    const findEdits = type === 'block' ? diffChangeBlock : diffByLine;

    const [oldEdits, newEdits] = changeBlocks.map(findEdits).reduce(
        ([oldEdits, newEdits], [currentOld, currentNew]) => [
            oldEdits.concat(currentOld),
            newEdits.concat(currentNew),
        ],
        [[], []]
    );

    return pickRanges(flatten(oldEdits), flatten(newEdits));
};
