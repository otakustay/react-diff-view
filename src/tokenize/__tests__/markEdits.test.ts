// @ts-nocheck ignore: third-party
import {basicHunk, multipleHunk} from '../../__tests__/cases';
import markEdits from '../markEdits';

describe('markEdits', () => {
    test('returns a function', () => {
        expect(typeof markEdits([], {})).toBe('function');
    });

    test('empty', () => {
        expect(markEdits([], {})([[{}], [{}]])).toEqual([[{}], [{}]]);
        expect(markEdits([], {type: 'line'})([[{}], [{}]])).toEqual([[{}], [{}]]);
    });

    test('withHunk', () => {
        expect(markEdits([basicHunk], {})([[], []])).toMatchSnapshot();
        expect(markEdits([multipleHunk], {})([[], []])).toMatchSnapshot();
        expect(markEdits([basicHunk], {type: 'line'})([[], []])).toMatchSnapshot();
        expect(markEdits([multipleHunk], {type: 'line'})([[], []])).toMatchSnapshot();
    });
});
