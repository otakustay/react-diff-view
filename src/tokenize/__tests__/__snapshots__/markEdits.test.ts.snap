// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markEdits withHunk 1`] = `
Array [
  Array [],
  Array [],
]
`;

exports[`markEdits withHunk 2`] = `
Array [
  Array [],
  Array [],
]
`;

exports[`markEdits withHunk 3`] = `
Array [
  Array [],
  Array [],
]
`;

exports[`markEdits withHunk 4`] = `
Array [
  Array [],
  Array [],
]
`;
